### DISCUSSION TITLE
HTMX history cache and Alpine template tags not working together

### DISCUSSION THREAD

**Discussion by gone** [alpinejs/alpine#2924] [7. Bugs (archived)]
I think this is potentially a feature request and not a bug:

I have a sample app (https://github.com/rwwiv/min_test_alpine_htmx)  that has a page that uses a template tag and x-for to do a little bit of dynamic rendering. If you open that page (either page2 or page3 in the app), move to another page and then hit back you will get errors in the console, and duplicate html.

 I am pretty sure what is happening is that when the back event occurs, htmx is restoring a cached html state (https://htmx.org/docs/#history)  which includes the prior alpine state, and also the template tag and alpine directives that created the state. The alpine directives re-run, and the old state is preserved but is no longer managed by alpine. 

Is there some way we can serialize the current state such that when we restore the cache alpine knows to pick that up rather than starting fresh? Especially if that includes data and not just html tags.

**Comment by gone** [comment]
I am thinking that the swapping mechanism htmx uses is the default one, and potentially I need to figure out how to use alpine.morph through https://htmx.org/extensions/alpine-morph/

**Reply by bmihelac** [reply]
@gone did you have any success with this problem? Having similar issue but no luck so far in finding the solution.

**Reply by gone** [reply]
no - I turned off the back cache

**Reply by quimdev19** [reply]
> no - I turned off the back cache

How did you disabled the HTMX cache? Changing the source code?

**Reply by gone** [reply]
https://github.com/Lightmatter/django-hydra/blob/develop/%7B%7Bcookiecutter.repo_name%7D%7D/frontend/js/main.ts#L23-L24

**Comment by Tim-Wils** [comment]
Have you seen this one?: https://dev.htmx.org/extensions/alpine-morph/

**Comment by ekwoka** [comment]
If Alpine-morph extension can't solve it, this is more of an HTMX issue. HTMX maintainers like Alpine, and want it to be compatible.

Alpine can't really do anything about it.

**Reply by hudon** [reply]
Why is it an htmx issue? It isn't doing anything fancy beyond swapping out \<body\>'s innerHTML with a snapshot. I could see other libraries doing similar things. Alpine's \<template x-for> logic could handle this without knowing about htmx.

**Reply by devwaseem** [reply]
I agree with @hudon. We need an API from alpine to serialize the existing state and restore it on demand. This could be a separate extension rather than being supported on the Alpine core. So that other libs needing this behavior can write an extension to support this particular use case.

**Comment by devwaseem** [comment]
I think I found the root cause of the problem. The template child elements are being saved to the cache, and when being restored from the cache, alpine loses the reference of the template's child element, thus considering them as a different element. I found a temporary workaround which is adding "x-from-template" attribute to the template's children. and once the HTMX restores the document, I just select and remove the elements which contain that attribute.

Example:
HTML before alpine init
```HTML
<template x-for="element in list">
    <li x-text="element.label" x-from-template></li>
</template>
```
After alpine processes the document, it turns into:

```HTML
<template x-for="element in list">...</template>
<li x-text="element.label" x-from-template></li>
<li x-text="element.label" x-from-template></li>
<li x-text="element.label" x-from-template></li>
...
```

But when this document is being stored to cache by htmx and when restored, alpine loses `<li>` tag references and considers them as a separate element.

To fix this issue, we can remove those `<li>` elements right after htmx restores the document from the cache
```HTML
<script>
    document.addEventListener('htmx:historyRestore', (evt) => {
        document.querySelectorAll('[x-from-template]').forEach((e) => e.remove())
    })
</script>
```

This will remove the orphaned template children and thus prevent errors.

Of course, this is a temporary workaround, I would like to know if we have a way from the alpine library to clean up template tags on demand or just re-attach them to the template reference. 

If we have any way to clean up template tags, we can create a htmx plugin to run the clean-up function every time the page is stored in to cache.

**Reply by drkane** [reply]
This was exactly what I needed -thanks!

**Reply by davemasters1984** [reply]
Thank you so much! Been pulling my hair out!

**Reply by aunjaffery** [reply]
exactly what i needed... also worked with x-if. thanks man. <3

**Reply by iggyzuk** [reply]
You can also use `htmx:beforeHistorySave` instead of `htmx:historyRestore`. This way you avoid adding those elements in the history entirely.

**Reply by OutOfBoundCats** [reply]
this worked for me thank you

**Comment by hudon** [comment]
For reference for the discussion, here are the htmx docs on the feature:
https://htmx.org/attributes/hx-push-url/ and https://htmx.org/attributes/hx-history-elt/
and here is the code for restoring the history cache: https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L2068

**Comment by calvary-tucson** [comment]
Possible solution that might be closer to the htmx way by preventing the content inside the template from being stored in htmx's history cache local storage object. This also allows the history cache to still work for other elements.

Use `hx-history="false"` on the child elements inside the template tag. - See hx-history documentation: https://htmx.org/attributes/hx-history/

Example usage for x-for:
```html
<template x-for="element in list">
    <li x-text="element.label" hx-history="false"></li>
</template>
```

Example usage for x-if:
```html
<template x-if="true">
    <li hx-history="false">Hello World</li>
</template>
```

**Reply by oselz** [reply]
`hx-history` on any element will disable history snapshotting for the entire page, not just the single element:
> Notes
> hx-history="false" can be present anywhere in the document to embargo the current page state from the history cache (i.e. even outside the element specified for the history snapshot [hx-history-elt](https://htmx.org/attributes/hx-history-elt/)).

**Comment by ekwoka** [comment]
The issue basically stems from HTMX just saving the document state as html, which alpine has no way to know that the element nearby is actually the contents of the template tag when the new template is evaluated

**Comment by danie-ramdhani** [comment]
Actually, this seems happen in livewire v3 too when using `wire:navigate` or `x-navigate`.
workaround: https://github.com/livewire/livewire/discussions/6849
failing test: https://github.com/livewire/livewire/pull/7034

**Reply by danie-ramdhani** [reply]
fixed! https://github.com/alpinejs/alpine/pull/4015
